app/scripts/lib/router.js

```diff
@@ -159,17 +159,43 @@ define(function (require, exports, module) {
     },
 
     /**
-     * Redirect the user to the best suitable OAuth flow
+     * Redirect the user to the best suitable OAuth flow.
      */
     redirectToBestOAuthChoice: function () {
-      var account = this.user.getChooserAccount();
+      var self = this;
       var route = '/oauth/signin';
 
-      if (account.isDefault()) {
-        route = '/oauth/signup';
-      }
+      var account = this.user.getChooserAccount();
 
-      return this.navigate(route, { replace: true, trigger: true });
+      // Attempt to get email address from relier
+      var email = this.broker.relier.get('email');
+
+      // Check whether email address corresponds to an existing account and
+      // navigate to correct login/signup page.
+      if (email) {
+        return account.checkAccountExistsByEmail(email)
+        .then(function (exists) {
+          if (! exists) {
+            route = '/oauth/signup';
```

Comment:
I feel like this whole function is useful outside of OAuth, the firstrun folks want the same functionality. .... .... ...